work/work.go

```diff
@@ -0,0 +1,191 @@
+package work
+
+import (
+	"context"
+	errs "errors"
+	"fmt"
+	"log"
+	"math/rand"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/heptio/workgroup"
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/stacktrace"
+
+	"github.com/elastic/hey-apm/out"
+)
+
+type Worker struct {
+	// not to be modified concurrently
+	workgroup.Group
+
+	RunTimeout             time.Duration
+	AgentFlushTimeout      time.Duration
+	ServerUrl              string
+	SecretToken            string
+	TransactionLimit       int
+	TransactionFrequency   time.Duration
+	MaxSpansPerTransaction int
+	MinSpansPerTransaction int
+	ErrorLimit             int
+	ErrorFrequency         time.Duration
+	MaxFramesPerError      int
+	MinFramesPerError      int
+}
+
+type Report struct {
+	TransactionsSent uint64
+	SpansSent        uint64
+	ErrorsSent       uint64
+	Start            time.Time
+	End              time.Time
+	RequestErrors    uint64
+}
+
+func (w *Worker) Work() (Report, error) {
+
+	logger := out.NewApmLogger(log.New(os.Stderr, "", log.Ldate|log.Ltime|log.Lshortfile))
+	if w.RunTimeout < time.Second {
+		return Report{}, errs.New("run timeout too short")
+	}
+
+	tracer := newTracer(logger, w.AgentFlushTimeout, w.SecretToken, w.ServerUrl, w.MaxSpansPerTransaction)
```

Comment:
how about breaking tracer into it's own module and passing one into work (either function or the worker struct) ?  That could help trim the worker struct down to just the work-related configuration and clean up the tracer api a bit.